### PR TITLE
feat: split fuse api for fs

### PR DESCRIPTION
**Pull request by IRONICBo**
# Background

At present, the storage interface exposed to the outside world is POSIX (fuse), in order to consider the user space to complete the transmission of storage data, consider providing a set of sdk for the business layer to efficiently read the data stored in the back-end without going through the kernel, you can refer to the form of the deployment of the following two main forms of discussion. 

# Core Modification

1. Migrate memfs to src/fs and rename to datenlordfs.
2. Split memfs to FuseFilesSystem and VirtualFs.
3. Move fuse params to fs_utils.
4. Remove metadata fuse operation.

# TODO

1. buffer trait, current implmentation is use BytesMut instead，we need to use common trait.
2. modify direntry to buffer

### CODE REVIEW

File: src/common/logger.rs [hunk 1]
```diff
@@ -62,8 +67,9 @@ pub fn init_logger(role: LogRole, level: Level) {
         .with_target("hyper", Level::WARN)
         .with_target("h2", Level::WARN)
         .with_target("tower", Level::WARN)
-        .with_target("datenlord::async_fuse::fuse", Level::INFO)
+        .with_target("datenlord::async_fuse::fuse", Level::ERROR)
```
Comment (frostyplanet) [inline src/common/logger.rs:70]: Do not change default log level

File: src/common/logger.rs [hunk 2]
```diff
@@ -62,8 +67,9 @@ pub fn init_logger(role: LogRole, level: Level) {
...
         .with_target("datenlord::metrics", Level::INFO)
+        .with_target("datenlordsdk", Level::DEBUG)
```
Comment (frostyplanet) [inline src/common/logger.rs:72]: Should be info

File: src/common/task_manager/gc.rs [hunk 3]
```diff
@@ -110,9 +110,11 @@ impl GcTask {
     /// Run the GC task, consume the task itself.
     #[allow(clippy::pattern_type_mismatch)] // for `tokio::select!`
     pub async fn run(mut self, token: CancellationToken) -> Self {
+        info!("GC task `{:?}` starts.", self.name);
```
Comment (frostyplanet) [inline src/common/task_manager/gc.rs:113]: These debug log has no use, should delete

File: src/common/task_manager/gc.rs [hunk 4]
```diff
@@ -110,9 +110,11 @@ impl GcTask {
...
         loop {
             select! {
                 res = self.rx.recv() => {
+                    info!("Received a new handle from `{:?}`.", self.name);
```
Comment (frostyplanet) [inline src/common/task_manager/gc.rs:117]: Should use debug

File: src/fs/virtualfs.rs [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    }
+
+    /// Open a file
+    async fn open(&self, uid: u32, gid: u32, ino: u64, flags: u32) -> DatenLordResult<u64>;
+
+    /// Read data with the given buffer, return current offset and the number of bytes read
+    async fn read(
+        &self,
+        ino: u64,
+        fh: u64,
```
Comment (frostyplanet) [inline src/fs/virtualfs.rs:116]: remove fh param

File: src/fs/virtualfs.rs [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+        offset: u64,
+        size: u32,
+        buf: &mut [u8],
+    ) -> DatenLordResult<usize>;
+
+    /// Write data
+    async fn write(
+        &self,
+        ino: u64,
+        fh: u64,
```
Comment (frostyplanet) [inline src/fs/virtualfs.rs:126]: remove fh param

File: src/fs/virtualfs.rs [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+        offset: i64,
+        data: &[u8],
+        flags: u32,
+    ) -> DatenLordResult<()>;
+
+    /// Flush method
+    async fn flush(&self, ino: u64, fh: u64, lock_owner: u64) -> DatenLordResult<()>;
```
Comment (frostyplanet) [inline src/fs/virtualfs.rs:133]: same

File: src/fs/virtualfs.rs [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+
+    /// Release an open file
+    async fn release(
+        &self,
+        ino: u64,
+        fh: u64,
+        flags: u32, // same as the open flags
+        lock_owner: u64,
+        flush: bool,
+    ) -> DatenLordResult<()>;
```
Comment (frostyplanet) [inline src/fs/virtualfs.rs:143]: same

File: src/fs/virtualfs.rs [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+
+    /// Synchronize file contents
+    async fn fsync(&self, ino: u64, fh: u64, datasync: bool) -> DatenLordResult<()>;
```
Comment (frostyplanet) [inline src/fs/virtualfs.rs:146]: same

File: src/new_storage/handle/reader.rs [hunk 10]
```diff
@@ -103,14 +104,25 @@ impl Reader {
                 let size: usize = slice.size.cast();
                 let end = offset + size;
                 let block_size = block.len();
+                error!(
```
Comment (frostyplanet) [inline src/new_storage/handle/reader.rs:107]: Remove this log, it's not error

File: src/new_storage/handle/reader.rs [hunk 11]
```diff
@@ -121,11 +133,13 @@ impl Reader {
...
 
     /// Close the reader and remove the accessed cache blocks.
+    #[allow(clippy::unwrap_used)]
+    #[allow(clippy::unused_self)]
     pub fn close(&self) {
-        let access_keys = self.access_keys.lock();
-        for key in access_keys.iter() {
-            self.cache.lock().remove(key);
-        }
+        // let access_keys = self.access_keys.lock();
```
Comment (frostyplanet) [inline src/new_storage/handle/reader.rs:139]: remove commented code

File: src/new_storage/handle/writer.rs [hunk 12]
```diff
@@ -147,8 +149,10 @@ async fn write_back_work(mut write_back_receiver: Receiver<Task>) {
     loop {
         tokio::select! {
             Some(task) = write_back_receiver.recv() => {
+                info!("Received a task: {:?}", task);
```
Comment (frostyplanet) [inline src/new_storage/handle/writer.rs:152]: Remove these info log

**Review by frostyplanet — CHANGES_REQUESTED** [review]

File: src/new_storage/storage_manager/manager.rs [hunk 13]
```diff
@@ -47,9 +47,16 @@ impl Storage for StorageManager {
     /// Reads data from a file specified by the file handle, starting at the
     /// given offset and reading up to `len` bytes.
     #[inline]
-    async fn read(&self, _ino: u64, fh: u64, offset: u64, len: usize) -> StorageResult<Vec<u8>> {
+    async fn read(
+        &self,
+        _ino: u64,
+        fh: u64,
```
Comment (frostyplanet) [inline src/new_storage/storage_manager/manager.rs:53]: remove fh param

File: src/new_storage/handle/writer.rs [hunk 14]
```diff
@@ -298,6 +304,7 @@ impl Writer {
                 .unwrap_or_else(|_| {
                     panic!("Should not send command to write back task when the task quits.");
                 });
+            info!("Write block {} back to the backend.", block_id);
```
Comment (frostyplanet) [inline src/new_storage/handle/writer.rs:307]: Remove these log

File: src/new_storage/handle/handle.rs [hunk 15]
```diff
@@ -302,8 +313,8 @@ mod tests {
         handles.add_handle(file_handle.clone());
         let buf = vec![b'1', b'2', b'3', b'4'];
         file_handle.write(0, &buf).await.unwrap();
-        let read_buf = file_handle.read(0, 4).await.unwrap();
-        assert_eq!(read_buf, buf);
-        file_handle.flush().await.unwrap();
+        // let read_buf = file_handle.read(0, 4).await.unwrap();
```
Comment (frostyplanet) [inline src/new_storage/handle/handle.rs:316]: fix this

File: src/storage/memory_cache/write_back_task.rs [hunk 16]
```diff
@@ -318,6 +318,8 @@ where
                 interval = tokio::time::interval(new_period);
             }
 
+            info!("Write back task is running.");
```
Comment (frostyplanet) [inline src/storage/memory_cache/write_back_task.rs:321]: remove this log

**Review by frostyplanet — CHANGES_REQUESTED** [review]

File: src/common/logger.rs (hunk 1 above)
Comment (IRONICBo) [reply src/common/logger.rs:70]: Done.

File: src/common/logger.rs (hunk 2 above)
Comment (IRONICBo) [reply src/common/logger.rs:72]: Done.

File: src/common/task_manager/gc.rs (hunk 3 above)
Comment (IRONICBo) [reply src/common/task_manager/gc.rs:113]: Done.

File: src/common/task_manager/gc.rs (hunk 4 above)
Comment (IRONICBo) [reply src/common/task_manager/gc.rs:117]: Done.

File: src/fs/virtualfs.rs (hunk 5 above)
Comment (IRONICBo) [reply src/fs/virtualfs.rs:116]: Ok, I have removed all fh params in fs.

File: src/fs/virtualfs.rs (hunk 6 above)
Comment (IRONICBo) [reply src/fs/virtualfs.rs:126]: Done.

File: src/fs/virtualfs.rs (hunk 7 above)
Comment (IRONICBo) [reply src/fs/virtualfs.rs:133]: Done.

File: src/fs/virtualfs.rs (hunk 8 above)
Comment (IRONICBo) [reply src/fs/virtualfs.rs:143]: Done.

File: src/fs/virtualfs.rs (hunk 9 above)
Comment (IRONICBo) [reply src/fs/virtualfs.rs:146]: Done.

File: src/new_storage/handle/reader.rs (hunk 10 above)
Comment (IRONICBo) [reply src/new_storage/handle/reader.rs:107]: OK, removed.

File: src/new_storage/handle/writer.rs (hunk 12 above)
Comment (IRONICBo) [reply src/new_storage/handle/writer.rs:152]: Done.

File: src/storage/memory_cache/write_back_task.rs (hunk 16 above)
Comment (IRONICBo) [reply src/storage/memory_cache/write_back_task.rs:321]: Done.

File: src/new_storage/handle/handle.rs (hunk 15 above)
Comment (IRONICBo) [reply src/new_storage/handle/handle.rs:316]: Done.

File: src/new_storage/handle/writer.rs (hunk 14 above)
Comment (IRONICBo) [reply src/new_storage/handle/writer.rs:307]: Done.

File: src/new_storage/storage_manager/manager.rs (hunk 13 above)
Comment (IRONICBo) [reply src/new_storage/storage_manager/manager.rs:53]: Done.

File: src/storage/memory_cache/write_back_task.rs (hunk 16 above)
Comment (IRONICBo) [reply src/storage/memory_cache/write_back_task.rs:321]: Done.

**Review by frostyplanet — APPROVED** [review]

**Pull request merged**